subprojects/core/src/main/java/org/gradle/invocation/DefaultGradle.java

```diff
@@ -629,4 +615,37 @@ public ClassLoaderScope getClassLoaderScope() {
     @Override
     @Inject
     public abstract PublicBuildPath getPublicBuildPath();
+
+    private DefaultGradleLifecycle instantiateGradleLifecycle() {
+        return services.get(ObjectFactory.class).newInstance(DefaultGradleLifecycle.class, this);
```

Comment:
❔ Do we use the factory here instead of a plain constructor call because we want the runtime decorated class to handle more variations of arguments for `beforeProject` calls in Groovy? Is there any other reason?

Consequently, does it mean that any new type in the Public API that accepts an isolated action has to be created through the object factory?

Reply:
Yes, it's the same story as for `Action<T>` taking methods. We need the runtime decoration so the `Closure` taking overloads are injected and setup the delegate correctly (so the `it` can be avoided).